R/utils.R

```diff
@@ -1838,3 +1838,36 @@ select_menu <- function(choices, title = NULL, msg = "Enter one or more numbers
     }
   }
 }
+
+#' @noRd
+isAppDir <- function(path) {
+
+  if (file.exists(file.path(path, "app.R")))
+    return(TRUE)
+
+  if (file.exists(file.path(path, "server.R"))
+      && file.exists(file.path(path, "ui.R")))
+    return(TRUE)
+
+  FALSE
+}
+
+# Borrowed from rprojroot which borrowed from devtools
+#' @noRd
+is_root <- function(path) {
+  identical(
+    normalizePath(path, winslash = "/"),
+    normalizePath(dirname(path), winslash = "/")
+  )
+}
+
+#' @noRd
+findEnclosingApp <- function(path = ".") {
+  path <- normalizePath(path, winslash = "/", mustWork = TRUE)
+  while (!is_root(path)) {
+    if (isAppDir(path))
+      return(path)
+    path <- dirname(path)
+  }
+  stop("Shiny app not found")
```

Comment:
It would be helpful to tell the user where it's looking:

```R
stop("Shiny app not found at ", orig_path, " or an ancestor directory.")
```

Where `orig_path` is the value passed to the function.